File: android/app/src/main/java/com/solace/ui/profile/ProfileCardCreationViewModel.kt
package com.solace.ui.profile

import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solace.data.user.UserRepository
import com.solace.ui.config.MoodColorKey
import com.solace.ui.config.UserProfile
import com.solace.ui.onboarding.OnboardingAnswers
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ProfileCreationState(
    val photoUri: Uri? = null,
    val isUploading: Boolean = false,
    val isSaved: Boolean = false,
    val error: String? = null,
)

@HiltViewModel
class ProfileCardCreationViewModel @Inject constructor(
    private val userRepo: UserRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(ProfileCreationState())
    val state: StateFlow<ProfileCreationState> = _state.asStateFlow()

    fun setPhoto(uri: Uri) = _state.update { it.copy(photoUri = uri) }

    fun clearError() = _state.update { it.copy(error = null) }

    fun publishCard(uid: String, answers: OnboardingAnswers) {
        viewModelScope.launch {
            _state.update { it.copy(isUploading = true, error = null) }
            try {
                val avatarUrl = _state.value.photoUri?.let { uri ->
                    userRepo.uploadAvatar(uid, uri).getOrThrow()
                }

                val profile = buildProfile(uid, answers, avatarUrl)
                userRepo.saveProfile(uid, profile).getOrThrow()
                _state.update { it.copy(isSaved = true, isUploading = false) }
            } catch (e: Exception) {
                _state.update { it.copy(isUploading = false, error = e.message) }
            }
        }
    }

    private fun buildProfile(uid: String, a: OnboardingAnswers, avatarUrl: String?): UserProfile {
        val moodEntry = DefaultMoodPalette.firstOrNull { it.key == a.moodColor }
        return UserProfile(
            userId = uid,
            username = a.username.ifBlank { "Anonymous" },
            avatarUrl = avatarUrl,
            moodColor = a.moodColor,
            moodLabel = moodEntry?.label ?: a.moodColor.name.replace('_', ' ').lowercase()
                .replaceFirstChar { it.uppercase() },
            bio = a.feelingSentence,
            vibes = a.vibes,
        )
    }
}

File: ios/Solace/ViewModels/ProfileCardCreationViewModel.swift
import SwiftUI
import PhotosUI

@MainActor
final class ProfileCardCreationViewModel: ObservableObject {

    @Published var selectedPhoto: PhotosPickerItem? = nil
    @Published var photoImage: UIImage? = nil
    @Published var isUploading: Bool = false
    @Published var isSaved: Bool = false
    @Published var error: String? = nil

    private let userRepo = UserRepository.shared

    func loadPhoto(from item: PhotosPickerItem?) async {
        guard let item else { return }
        if let data = try? await item.loadTransferable(type: Data.self),
           let img = UIImage(data: data) {
            photoImage = img
        }
    }

    func publishCard(uid: String, answers: OnboardingAnswers) async {
        isUploading = true
        error = nil
        do {
            let avatarUrl: String?
            if let img = photoImage, let data = img.jpegData(compressionQuality: 0.8) {
                avatarUrl = try await userRepo.uploadAvatar(uid: uid, imageData: data)
            } else {
                avatarUrl = nil
            }

            let profile = buildProfile(uid: uid, answers: answers, avatarUrl: avatarUrl)
            try await userRepo.saveProfile(uid: uid, profile: profile)
            isSaved = true
        } catch {
            self.error = error.localizedDescription
        }
        isUploading = false
    }

    private func buildProfile(uid: String, answers: OnboardingAnswers, avatarUrl: String?) -> UserProfile {
        let moodEntry = defaultMoodPalette.first { $0.key == answers.moodColor }
        return UserProfile(
            id: uid,
            username: answers.username.isEmpty ? "Anonymous" : answers.username,
            avatarUrl: avatarUrl,
            moodColor: answers.moodColor,
            moodLabel: moodEntry?.label ?? answers.moodColor.rawValue.replacingOccurrences(of: "_", with: " ").capitalized,
            bio: answers.feelingSentence,
            vibes: answers.vibes
        )
    }
}
